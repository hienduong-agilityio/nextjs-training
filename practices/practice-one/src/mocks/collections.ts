import { Category } from '@/types/category';

type HotDeal = {
  name: Category | 'all';
  count: number;
};

export const HOT_DEALS: HotDeal[] = [
  { name: 'all', count: 2 },
  { name: 'beauty', count: 2 },
  { name: 'fragrances', count: 48 },
  { name: 'furniture', count: 95 },
  { name: 'groceries', count: 23 },
];

export const BRANDS = [
  { name: 'Nike', count: 99 },
  { name: 'Adidas', count: 99 },
  { name: 'Siemens', count: 99 },
];

export const FILTER_GROUP = [{ title: 'Hot Deals', items: HOT_DEALS }];
