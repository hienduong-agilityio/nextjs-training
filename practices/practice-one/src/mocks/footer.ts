import { ROUTE } from '@/constants';

export const FOOTER_NAVIGATION = [
  {
    title: 'Information',
    items: [
      { name: 'About Us', url: ROUTE.ABOUT_US },
      { name: 'Information', url: ROUTE.INFORMATION },
      { name: 'Privacy Policy', url: ROUTE.PRIVACY_POLICY },
      { name: 'Terms & Conditions', url: ROUTE.TERM_AND_CONDITIONS },
    ],
  },
  {
    title: 'Service',
    items: [
      { name: 'About Us', url: ROUTE.ABOUT_US },
      { name: 'Information', url: ROUTE.INFORMATION },
      { name: 'Privacy Policy', url: ROUTE.PRIVACY_POLICY },
      { name: 'Terms & Conditions', url: ROUTE.TERM_AND_CONDITIONS },
    ],
  },
  {
    title: 'My Account',
    items: [
      { name: 'About Us', url: ROUTE.ABOUT_US },
      { name: 'Information', url: ROUTE.INFORMATION },
      { name: 'Privacy Policy', url: ROUTE.PRIVACY_POLICY },
      { name: 'Terms & Conditions', url: ROUTE.TERM_AND_CONDITIONS },
    ],
  },
  {
    title: 'Our Offers',
    items: [
      { name: 'About Us', url: ROUTE.ABOUT_US },
      { name: 'Information', url: ROUTE.INFORMATION },
      { name: 'Privacy Policy', url: ROUTE.PRIVACY_POLICY },
      { name: 'Terms & Conditions', url: ROUTE.TERM_AND_CONDITIONS },
    ],
  },
];
