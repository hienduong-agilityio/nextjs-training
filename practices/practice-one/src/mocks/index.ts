export * from './product';
export * from './service';
export * from './footer';
export * from './collections';
