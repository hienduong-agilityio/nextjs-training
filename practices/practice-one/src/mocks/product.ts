// Types
import type { Category } from '@/types';
import type { ICartItem, IProductProps } from '@/interfaces';

export const CATEGORIES: (Category | 'all')[] = [
  'all',
  'beauty',
  'fragrances',
  'furniture',
  'groceries',
];

export const PRODUCTS_DATA: IProductProps[] = [
  {
    id: '1',
    name: 'Sneakers',
    images: ['/images/product-mock.png'],
    price: 99,
    originalPrice: '$129',
    discount: '23%',
    rating: 5,
    category: 'fragrances',
  },
  {
    id: '2',
    name: 'Leather Shoulder Bag',
    images: ['/images/product-mock.png'],
    price: 299.43,
    originalPrice: '$534.33',
    discount: '44%',
    rating: 5,
    category: 'beauty',
  },
  {
    id: '3',
    name: 'Nike Air Max 270 React',
    images: ['/images/product-mock.png'],
    price: 299.43,
    originalPrice: '$534.33',
    discount: '44%',
    rating: 5,
    category: 'fragrances',
    label: 'HOT',
  },
  {
    id: '4',
    name: 'Classic Leather Belt',
    images: ['/images/product-mock.png'],
    price: 199.99,
    originalPrice: '$249.99',
    discount: '20%',
    rating: 4,
    category: 'furniture',
  },
  {
    id: '5',
    name: 'Aviator Sunglasses',
    images: ['/images/product-mock.png'],
    price: 159.99,
    originalPrice: '$199.99',
    discount: '20%',
    rating: 4,
    category: 'groceries',
  },
  {
    id: '6',
    name: 'Round Frame Sunglasses',
    images: ['/images/product-mock.png'],
    price: 139.99,
    originalPrice: '$179.99',
    discount: '22%',
    rating: 4,
    category: 'groceries',
  },
];

export const CART_DATA: ICartItem[] = [
  {
    id: '2',
    title: 'Eyeshadow Palette with Mirror',
    price: 19.99,
    quantity: 3,
    total: 59.97,
    discountPercentage: 5.5,
    discountedTotal: 56.67164999999999,
    thumbnail:
      'https://cdn.dummyjson.com/products/images/beauty/Eyeshadow%20Palette%20with%20Mirror/1.png',
  },
  {
    id: '1',
    title: 'Essence Mascara Lash Princess',
    price: 9.99,
    quantity: 1,
    total: 9.99,
    discountPercentage: 7.17,
    discountedTotal: 9.273717,
    thumbnail:
      'https://cdn.dummyjson.com/products/images/beauty/Essence%20Mascara%20Lash%20Princess/1.png',
  },
  {
    id: '3',
    title: 'Powder Canister',
    price: 14.99,
    quantity: 2,
    total: 29.98,
    discountPercentage: 18.14,
    discountedTotal: 24.541628,
    thumbnail:
      'https://cdn.dummyjson.com/products/images/beauty/Powder%20Canister/1.png',
  },
];
