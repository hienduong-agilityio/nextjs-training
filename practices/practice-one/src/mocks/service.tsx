import { RefundIcon, ShippingIcon, SupportIcon } from '@/icons';

export const serviceData = [
  {
    title: 'FREE SHIPPING',
    details:
      'Lorem Ipsum is simply dummy text of the printing and typesetting industry.',
    icon: <ShippingIcon size={100} className="w-24 h-24" />,
  },
  {
    title: '100% REFUND',
    details:
      'Lorem Ipsum is simply dummy text of the printing and typesetting industry.',
    icon: <RefundIcon size={100} />,
  },
  {
    title: 'SUPPORT 24/7',
    details:
      'Lorem Ipsum is simply dummy text of the printing and typesetting industry.',
    icon: <SupportIcon size={100} />,
  },
];

export const SERVICE_CARD = {
  default: {
    icon: 'ðŸ”§',
    title: 'Default Service',
    details:
      'Lorem Ipsum is simply dummy text of the printing and typesetting industry.',
  },
  customIcon: {
    icon: 'ðŸš€',
    title: 'Rocket Service',
    details:
      'Lorem Ipsum is simply dummy text of the printing and typesetting industry.',
  },
  withoutIcon: {
    title: 'Icon-Free Service',
    details:
      'Lorem Ipsum is simply dummy text of the printing and typesetting industry.',
  },
};
