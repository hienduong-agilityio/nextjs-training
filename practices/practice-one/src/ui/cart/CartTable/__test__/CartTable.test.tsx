import { render, screen } from '@testing-library/react';

// Types
import type { ICartItem } from '@/interfaces';

// UI
import { CartTable } from '../index';

// Mocks
import { CART_DATA } from '@/mocks';

jest.mock('@/ui', () => ({
  EmptyCart: jest.fn(() => <div data-testid="empty-cart">Empty Cart</div>),
  CartItemRow: jest.fn(({ id }) => (
    <tr data-testid="cart-item-row">
      <td>{id}</td>
    </tr>
  )),
}));

describe('CartTable', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  const renderCartTable = (products: ICartItem[]) => {
    render(<CartTable products={products} />);
  };

  const cartTableHeaders = ['Product', 'Price', 'Qty', 'Unit Price'];

  it('renders the empty cart state when products are empty', () => {
    renderCartTable([]);
    expect(screen.getByTestId('empty-cart')).toBeInTheDocument();
  });

  it('renders the cart table headers correctly', () => {
    renderCartTable(CART_DATA);

    cartTableHeaders.forEach((header) => {
      expect(screen.getByText(header)).toBeInTheDocument();
    });
  });

  it('renders CartItemRow for each product', () => {
    renderCartTable(CART_DATA);

    const cartItemRows = screen.getAllByTestId('cart-item-row');
    expect(cartItemRows).toHaveLength(CART_DATA.length);

    CART_DATA.forEach((product) => {
      expect(screen.getByText(product.id)).toBeInTheDocument();
    });
  });

  it('renders the table structure correctly', () => {
    renderCartTable(CART_DATA);

    expect(
      screen.getByRole('table', { name: 'Cart table' }),
    ).toBeInTheDocument();

    expect(
      screen.getByRole('rowgroup', { name: 'Cart table headers' }),
    ).toBeInTheDocument();
    expect(
      screen.getByRole('rowgroup', { name: 'Cart table body' }),
    ).toBeInTheDocument();

    const headerCells = screen.getAllByRole('columnheader');
    expect(headerCells).toHaveLength(cartTableHeaders.length);
    expect(headerCells.map((cell) => cell.textContent)).toEqual(
      cartTableHeaders,
    );
  });

  it('matches the snapshot', () => {
    const { asFragment } = render(<CartTable products={CART_DATA} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
