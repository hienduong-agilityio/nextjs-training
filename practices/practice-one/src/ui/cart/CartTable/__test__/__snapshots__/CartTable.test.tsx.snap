// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartTable matches the snapshot 1`] = `
<DocumentFragment>
  <section
    class="overflow-x-auto"
  >
    <table
      aria-label="Cart table"
      class="w-full text-left border-collapse table-auto"
    >
      <thead
        aria-label="Cart table headers"
        class="text-sm uppercase border-b-2 border-secondary-100"
      >
        <tr>
          <th
            class="text-xl font-medium whitespace-nowrap px-16 py-3"
          >
            Product
          </th>
          <th
            class="text-xl font-medium whitespace-nowrap px-4"
          >
            Price
          </th>
          <th
            class="text-xl font-medium whitespace-nowrap px-4"
          >
            Qty
          </th>
          <th
            class="text-xl font-medium whitespace-nowrap px-4"
          >
            Unit Price
          </th>
        </tr>
      </thead>
      <tbody
        aria-label="Cart table body"
        class="divide-y-2 border-b-2 border-secondary-100 divide-secondary-100"
      >
        <tr
          data-testid="cart-item-row"
        >
          <td>
            2
          </td>
        </tr>
        <tr
          data-testid="cart-item-row"
        >
          <td>
            1
          </td>
        </tr>
        <tr
          data-testid="cart-item-row"
        >
          <td>
            3
          </td>
        </tr>
      </tbody>
    </table>
  </section>
</DocumentFragment>
`;
